### PR TITLE
Move Active Authentication Mode into AutomaticAuthenticationHandler

### PR DESCRIPTION
**Pull request by HaoK**
- Rename AuthenticationMode to AutomaticAuthentication bool on AutomaticAuthenticationOptions
- Split old active behavior into a new base handler class : `AutomaticAuthenticationHandler` which derives from `AuthenticationHandler<AutomaticAuthenticationOptions>`
- Remove SecurityHelper LookupChallenge and added two new virtuals to handle the old active authentication checks (which work when no scheme is specified)
- Also attempted to address 401-403 with this change (needs more unit tests here)

cc @lodejard @Praburaj @blowdart @Tratcher

### CODE REVIEW

File: Security.sln [hunk 1]
```diff
@@ -60,6 +60,8 @@ Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.AspNet.Authorizat
 EndProject
 Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.AspNet.Authorization", "src\Microsoft.AspNet.Authorization\Microsoft.AspNet.Authorization.kproj", "{6AB3E514-5894-4131-9399-DC7D5284ADDB}"
 EndProject
+Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.Framework.WebEncoders", "..\HttpAbstractions\src\Microsoft.Framework.WebEncoders\Microsoft.Framework.WebEncoders.kproj", "{DD2CE416-765E-4000-A03E-C2FF165DA1B6}"
```
Comment (Tratcher) [inline Security.sln]: I assume you didn't mean to check in this relative reference.

Comment (HaoK) [inline Security.sln]: Yeah I'll revert the global.json and revert all the sln changes as part of the final squash/merge, they won't make it back to dev

File: src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs [hunk 2]
```diff
@@ -386,7 +381,16 @@ public virtual void Challenge(IChallengeContext context)
         /// <returns></returns>
         protected virtual Task ApplyResponseChallengeAsync()
         {
-            ApplyResponseChallenge();
+            // If authenticate was called and the the status is still 401, authZ failed so set 403 and stop
+            // REVIEW: Does this need to ensure that there's the 401 is challenge for this auth type?
+            if (ShouldConvertChallengeToForbidden())
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs]: All of this logic should be moved to AutomaticAuthenticationHandler, that's the only place where it should ever be true.

Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs]: Actually now that I remember some of our earlier conversations, this was supposed to move back down into CookieAuth & BearerAuth's ApplyResponseChallenge because CookieAuth doesn't want to issue 403s, it wants to redirect to an AccessDenied page similar to how it handles challenges.

Comment (HaoK) [inline src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs]: Okay I'll move the logic and also add a test verifying that Cookie/Bearer don't turn 401's into 403s if they weren't asked to produce a ticket

**Comment by HaoK** [comment]
Updated PR, moved ShouldConvertChallengeToForbidden to AutomaticHandler, added a test for cookie/bearer to ensure they don't touch 401's if they weren't authenticated.

File: src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs [hunk 3]
```diff
@@ -334,7 +314,7 @@ protected virtual Task ApplyResponseGrantAsync()
 
         public virtual void SignIn(ISignInContext context)
         {
-            SignInIdentityContext = new SignInIdentityContext(context.Principal, new AuthenticationProperties(context.Properties));
+            SignInContext = new SignInContext(context.Principal, new AuthenticationProperties(context.Properties));
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs]: Shouldn't SignIn and SignOut also check for an AuthScheme match before doing anything here?

File: src/Microsoft.AspNet.Authentication/AutomaticAuthenticationHandler.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                (authenticationSchemes == null || !authenticationSchemes.Any());
+        }
+
+        /// <summary>
+        /// Override this method to deal with 401 challenge concerns, if an authentication scheme in question
+        /// deals an authentication interaction as part of it's request flow. (like adding a response header, or
+        /// changing the 401 result to 302 of a login page or external sign-in location.)
+        /// </summary>
+        /// <returns></returns>
+        protected override Task ApplyResponseChallengeAsync()
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication/AutomaticAuthenticationHandler.cs]: This method is overridden in both Cookies and Bearer, so I don't think there's much use in even implementing it here.

File: src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs [hunk 5]
```diff
@@ -329,8 +328,8 @@ protected override void ApplyResponseChallenge()
                 return;
             }
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs]: We need to implement the Access Denied scenario here as a mirror of Options.LoginPath.

File: src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs (hunk 3 above)
Comment (HaoK) [inline src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs]: Doh yeah. I will add some tests verifying the negative case too

File: src/Microsoft.AspNet.Authentication/AutomaticAuthenticationHandler.cs (hunk 4 above)
Comment (HaoK) [inline src/Microsoft.AspNet.Authentication/AutomaticAuthenticationHandler.cs]: Well its a bit confusing, cookies is overriding the sync version only, it would have to override the async version as well then, so this method is whats being used for cookies to handle the 401-403

File: src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs (hunk 5 above)
Comment (HaoK) [inline src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs]: Filed https://github.com/aspnet/Security/issues/166 to deal with this in a different PR

File: src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs (hunk 3 above)
Comment (HaoK) [inline src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs]: Added tests verifying that exceptions are thrown (rather than sign in/sign out happening) when wrong auth type is specified

**Comment by HaoK** [comment]
Merged 775eb5ece452e10a27aa166f52b58a6c0b615d18

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenidConnectAuthenticationHandler.cs [hunk 6]
```diff
@@ -566,7 +566,7 @@ public override Task<bool> InvokeAsync()
             {
                 if (ticket.Principal != null)
                 {
-                    Request.HttpContext.Response.SignIn(ticket.Properties, ticket.Principal.Identities);
+                    Request.HttpContext.Response.SignIn(ticket.AuthenticationScheme, ticket.Principal, ticket.Properties);
```
Comment (PinpointTownes) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenidConnectAuthenticationHandler.cs]: An invalid scheme is used here and makes the OIDC client middleware totally buggy because it cannot delegate the principal persistence to the cookie middleware.

Use __Option.SignInScheme__ instead.

Comment (HaoK) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenidConnectAuthenticationHandler.cs]: Ah thanks, saved me from having to track this down today!

Comment (PinpointTownes) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenidConnectAuthenticationHandler.cs]: I opened a ticket for that (https://github.com/aspnet/Security/issues/167) and submitted a PR (https://github.com/aspnet/Security/pull/168) that fixes it :smile:
